partiql-conformance-tests/build.rs

```diff
@@ -0,0 +1,296 @@
+/// TODO: add additional fields for further error-checking (e.g. error code)
+struct ParseFailTestCase {
+    test_name: String,
+    statement: String,
+    namespace: Option<String>,
+}
+
+impl TestFunction for ParseFailTestCase {
+    /// Creates a [`codegen::Function`] that checks the parsing of `statement` results in a parsing
+    /// error.
+    fn to_test_function(&self) -> Function {
+        let complete_test_name = match &self.namespace {
+            Some(ns) => format!("{}_{}", ns, &self.test_name),
+            None => self.test_name.to_owned(),
+        };
+        let mut generated_fun = Function::new(complete_test_name.replace(' ', "_").as_str());
+        generated_fun
+            .attr("test")
+            .line(format!(
+                "let parse_result = lalr_parse(\"{}\");",
+                &self.statement
+            ))
+            .line("assert!(parse_result.is_err());");
+        generated_fun
+    }
+}
+
+/// Parses the given .ion file `path` and creates a new [`codegen::Scope`] with a
+/// [`codegen::Function`] for each test in `path`.
+///
+/// Assumes the given tests in `path` all parse without errors.
+fn create_pass_parse_scope(path: &Path) -> Scope {
+    let all_data = fs::read(path).unwrap();
+    let all_ion_data = element_reader().read_all(&all_data).unwrap();
+
+    let mut scope = Scope::new();
+
+    for ion_data in all_ion_data {
+        match ion_data.ion_type() {
+            IonType::Struct => {
+                let test_struct = ion_data.as_struct().unwrap();
+                let test_name = test_struct
+                    .get("name")
+                    .unwrap()
+                    .as_str()
+                    .unwrap()
+                    .to_owned();
+                let statement = test_struct
+                    .get("statement")
+                    .unwrap()
+                    .as_str()
+                    .unwrap()
+                    .to_owned();
+
+                let test_case = ParsePassTestCase {
+                    test_name,
+                    statement,
+                    namespace: None,
+                };
+                scope.push_fn(test_case.to_test_function());
+            }
+            IonType::List => {
+                let test_list = ion_data.as_sequence().unwrap();
+                for tc in test_list.iter() {
+                    match tc.ion_type() {
+                        IonType::Struct => {
+                            let namespace = Some(
+                                ion_data
+                                    .annotations()
+                                    .next()
+                                    .unwrap()
+                                    .text()
+                                    .unwrap()
+                                    .to_owned(),
+                            );
+
+                            let test_struct = tc.as_struct().unwrap();
```

Comment:
This duplicates the above parsing. I think it would be better to model this a simply recursion of ion lists (representing namespace levels that get turned into scopes) with structs at the bottom (representing the test function that gets added to the closest enclosing scope)